### ISSUE TITLE
Tracking issue: 32bit x86 targets lose NaN payload in return values

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#115567]
On x86 (32bit) targets,  returning a floating-point number by-value can change the bits of their NaN payloads. This is caused by using x87 registers to pass return values, as mandated by the ABI for this target.

This is a known and long-standing problem, and very hard to fix. The purpose of this issue mostly is to document its existence and to give it a URL that can be referenced.

A complete fix is not possible since the ABI for `extern "C"` functions is not up for change. However we could change the default "Rust" ABI to not use the x87 registers ever if SSE is available. That would make the problem much less likely to appear in practice.

Related/prior issues:
- https://github.com/rust-lang/rust/issues/73288

**Comment by Diggsey** [comment]
It looks like the 80-bit float load/store operations don't modify the bits, so in principle one could fix this by performing the conversion from 32-bit or 64-bit float to 80-bit float (and back) in software, although I expect that would be prohibitively expensive.

**Comment by RalfJung** [comment]
Could we do a quick check whether the float is a NaN or not, and only do the software conversion for NaNs? That should be a lot faster I hope.

**Comment by Diggsey** [comment]
Came up with this code for software conversion:
```c

typedef union {
    long double ld;
    struct {
        unsigned long long ull;
        unsigned short us;
    };
} ld_t;

typedef union {
    double d;
    unsigned long long ull;
} d_t;

long double identity(double x) {
    d_t y;
    y.d = x;
    ld_t z;
    z.ull = (y.ull << 11) & 0x7FFFFFFFFFFFFFFFull | 0x8000000000000000ull;
    unsigned short e = (y.ull >> 52) & 0x7FF;
    z.us = ((e == 0x7FF) ? 0x7FFF : e + 0x3C00) | ((y.ull >> 48) & 0x8000);
    return z.ld;
}
```

```asm
identity:
        sub     esp, 28
        movq    xmm1, QWORD PTR [esp+32]
        movdqa  xmm0, xmm1
        psllq   xmm1, 11
        psrlq   xmm0, 32
        movd    eax, xmm0
        movdqa  xmm0, XMMWORD PTR .LC0
        mov     edx, eax
        shr     edx, 20
        por     xmm1, xmm0
        and     dx, 2047
        movq    QWORD PTR [esp], xmm1
        cmp     dx, 2047
        lea     ecx, [edx+15360]
        mov     edx, 32767
        cmovne  edx, ecx
        shr     eax, 16
        and     ax, -32768
        or      eax, edx
        mov     WORD PTR [esp+8], ax
        fld     TBYTE PTR [esp]
        add     esp, 28
        ret
.LC0:
        .long   0
        .long   -2147483648
        .long   0
        .long   0
```

Or, special casing NaNs:
```c
#include "math.h"

typedef union {
    long double ld;
    struct {
        unsigned long long ull;
        unsigned short us;
    };
} ld_t;

typedef union {
    double d;
    unsigned long long ull;
} d_t;

long double identity(double x) {
    if (!isnan(x)) {
        return x;
    }
    d_t y;
    y.d = x;
    ld_t z;
    z.ull = (y.ull << 11) & 0x7FFFFFFFFFFFFFFFull | 0x8000000000000000ull;
    z.us = (y.ull >> 48) | 0x7FFF;
    return z.ld;
}
```

```asm
identity:
        sub     esp, 44
        fld     QWORD PTR [esp+48]
        fst     QWORD PTR [esp+8]
        fucomi  st, st(0)
        jp      .L2
        add     esp, 44
        ret
.L2:
        fstp    st(0)
        mov     eax, DWORD PTR [esp+8]
        mov     edx, DWORD PTR [esp+12]
        movd    xmm0, eax
        movd    xmm1, edx
        mov     eax, edx
        punpckldq       xmm0, xmm1
        shr     eax, 16
        movdqa  xmm1, XMMWORD PTR .LC0
        psllq   xmm0, 11
        or      ax, 32767
        por     xmm0, xmm1
        mov     WORD PTR [esp+24], ax
        movq    QWORD PTR [esp+16], xmm0
        fld     TBYTE PTR [esp+16]
        add     esp, 44
        ret
.LC0:
        .long   0
        .long   -2147483648
        .long   0
        .long   0```

**Comment by RalfJung** [comment]
Nice. :)

So, I wonder -- what would it take to actually use this when returning floats or doubles? Do we have to patch LLVM or can the codegen backend do this?

**Comment by RalfJung** [comment]
Btw, can someone explain why this only affects return values? How are float arguments passed?

**Comment by Diggsey** [comment]
> So, I wonder -- what would it take to actually use this when returning floats or doubles? Do we have to patch LLVM or can the codegen backend do this?

No idea :)

> Btw, can someone explain why this only affects return values? How are float arguments passed?

Arguments on x86 (with the C calling convention) are always passed on the stack regardless of type. So the problem still exists as soon as you `fld` the value, but it's not inherent to the calling convention.

**Comment by comex** [comment]
By the same token, for non-`extern "C"` functions, floating point returns could be changed to go on the stack, even without SSE.

**Comment by RalfJung** [comment]
[Here](https://github.com/llvm/llvm-project/issues/66803#issuecomment-1856428859) there is a claim that this issue only affects *signaling* NaN return values. Is that true? The example given [here](https://github.com/rust-lang/rust/issues/73288) does seem to be a signaling NaN.

**Comment by jerrywrice** [comment]
Until a reasonable fix is chosen and incorporated, shouldn't there be a priority effort to patch these target's generated code so when this specific float value corruption occurs at run-time, either a panic is generated or a run-time error message logged? Or maybe this run-time safety check patch should be added via a feature or config option, where later it's removed when the proper fix is incorporated. Otherwise, unless I'm misinterpreting, it sounds as if broken (unsound?) code is currently allowed 'in the wild' (even if in rare circumstances)? Or is this so rare that no-one who it impacts is concerned? If this is the case, then I hope that at the language documentation (and relevant crate documentation) prominently mentions this issue so it's not easily overlooked by anyone developing new code for these targets. granted that new code for these targets is probably itself a rare occurrence).

**Comment by RalfJung** [comment]
> Otherwise, unless I'm misinterpreting, it sounds as if broken (unsound?) code is currently allowed 'in the wild' (even if in rare circumstances)? Or is this so rare that no-one who it impacts is concerned?

Yes, this is a miscompilation bug. It looks like it just doesn't affect enough people to motivate anyone to fix it. x86-32 is a fairly niche target these days, and signaling NaNs are quite rare as well, so very, very few people care about the combination of those. Combine that with the fact that very few people have the expertise to fix a problem like this, and those people typically have higher-impact issues to worry about, and you get a perfect recipe for problems like this flying under the radar.

>  If this is the case, then I hope that the relevant language documentation and relevant crate documentation (not just the target Tier documentation)

This is a bug in how a target implements the language, so the target docs seem like the primary place to mention issues like this. I'm not sure where in the language docs it'd make sense to put this -- I guess we could have a section on [this page](https://doc.rust-lang.org/nightly/std/primitive.f32.html) summarizing target-specific floating point issues?
